### ISSUE TITLE
x86-64 Compatability

### ISSUE THREAD

**Issue by Ericson2314** [Ericson2314/Voxlap#17]

**Comment by iamgreaser** [comment]
Very important because 64-bit is awesome and you should all be porting everything to 64-bit.

With that said, porting the SSE stuff to compiler intrinsics would make this part _much_ easier.

**Comment by aponigricon** [comment]
Approved

**Comment by Ericson2314** [comment]
Only v5.asm NEEDs that for x86-64, for the rest the C equivalents will do as a stop gap. On the other hand fixing subtle errors in the C types may be harder than the assembly

**Comment by Lensman** [comment]
Seems worthwhile, especially converting to intrin.

**Comment by Ericson2314** [comment]
In the end, I certainly want intrinsics and C working on x86-64

**Comment by Lensman** [comment]
Will have to be careful of __m64 datatypes. Probably have to convert them to __128 equivalents, which may mean wasted ops unless algorithms are also converted to take advantage of extra bandwidth eg;

static inline void mmxcolorsub(long _color){
 *color = _m_to_int(_mm_subs_pu8(_m_from_int(_color), _m_from_int(flashbrival)));
}

**Comment by Ericson2314** [comment]
What exactly is the problem with __m64?

**Comment by Lensman** [comment]
Nothing is wrong with it, but intrinsics that support that datatype are not supported on x86_64, so it will mean hand coded assembly. I have converted hrendzsse to intrinsics, and surprisingly it runs faster than the original code. Go figure ;)
It's all 128 bit xxm registers now, and on x86_64 there are 8 more to play with too. Thought this might be agood place to tackle in my spare time as it covers 6 asm functions;
hrendzsse/3dn, hrendsse/3dn vrendsse/3dn etc. I haven't finished the fog versions, but it's trivial to do so. This should collapse 8 functions into one more manageble one. I also had a play with mem mapped files, and now have voxlap working with 4096x4096x256 faster than the original. I'd say that is some progress.

**Comment by Ericson2314** [comment]
Wow, I'll be back working on Voxlap about a week, but it looks like you might have it done before then!

**Comment by iamgreaser** [comment]
The point you made about __m64 not working on x86-64 is incorrect - it DOES work, at least in GCC.
But yeah, if you're willing to do SSE2 support, then using an __m128 version of the code should be a bit faster.

**Comment by aponigricon** [comment]
The point _I_ made?

**Comment by Ericson2314** [comment]
Lensman I think.

**Comment by aponigricon** [comment]
Oh, well... my computer is broken... and I wont get my fingers on it's keyboard before at least two weeks, sooo...
What have you guys actually been doing?

**Comment by Ericson2314** [comment]
https://github.com/Ericson2314/henet
https://github.com/Ericson2314/hasnip

**Comment by aponigricon** [comment]
Haskell? Dude... NO... please... NOOOO...

**Comment by Ericson2314** [comment]
Yup. I may use Voxlap, but everything else will be Haskell. Seriously C is a shitty language. Rust or ATS is much better for manual memory.
